File: cocos/asset/assets/texture-2d-array.ts
import { EDITOR, TEST } from 'internal:constants';
import { cclegacy, js } from '../../core';
import { PresumedGFXTextureInfo, PresumedGFXTextureViewInfo, SimpleTexture } from './simple-texture';

import { ccclass, serializable, type } from 'cc.decorator';
import { PixelFormat } from './asset-enum';
import { ImageAsset } from './image-asset';
import { TextureInfo, TextureType, TextureViewInfo } from '../../gfx';
import { Texture2D } from './texture-2d';
import { max } from '../../core/math/bits';

/**
 * @en The create information for [[Texture2DArray]].
 * @zh 用来创建2D数组贴图的信息。
 */
export interface ITexture2DArrayCreateInfo {

    /**
     * @en The pixel width.
     * @zh 像素宽度。
     */
    width: number;

    /**
     * @en The pixel height.
     * @zh 像素高度。
     */
    height: number;

    /**
     * @en The pixel format.
     * @zh 像素格式。
     * @default PixelFormat.RGBA8888
     */
    format?: PixelFormat;

    /**
     * @en The mipmap level count.
     * @zh mipmap 层级。
     * @default 1
     */
    mipmapLevel?: number;

    /**
     * @en The selected base mipmap level.
     * @zh 选择使用的最小 mipmap 层级。
     * @default 1
     */
    baseLevel?: number;

    /**
     * @en The selected maximum mipmap level.
     * @zh 选择使用的最大 mipmap 层级。
     * @default 1000
     */
    maxLevel?: number;

}

type ITexture2DArrayMipmaps = ImageAsset[];

@ccclass('cc.Texture2DArray')
export class Texture2DArray extends SimpleTexture {

    // @serializable
    @type([[ImageAsset]])
    public _mipmaps: ITexture2DArrayMipmaps[] = [];

    private _generatedMipmaps: ITexture2DArrayMipmaps[] = [];

    get mipmaps(): ITexture2DArrayMipmaps[] {
        return this._mipmaps;
    }

    set mipmaps(value: ITexture2DArrayMipmaps[]) {
        this._mipmaps = value;

        const mipmaps: ImageAsset[][] = [];

        if (value.length === 1) {
            const mipmapsSlices = value[0];
            const sliceMipmaps: ImageAsset[][] = [];
            let maxMipmapLevel: number = 0;

            for (let n = 0; n < mipmapsSlices.length; n++) {
                sliceMipmaps[n] = mipmapsSlices[n].extractMipmaps();
                maxMipmapLevel = max(maxMipmapLevel, sliceMipmaps[n].length);
            }

            for (let n = 0; n < mipmapsSlices.length; n++) {
                if (sliceMipmaps[n].length != maxMipmapLevel) {
                    console.error('The number of mipmaps of each slice is different.');
                    this._setMipmapParams([]);
                    return;
                }
            }

            for (let i = 0; i < maxMipmapLevel; i++) {
                for (let n = 0; n < sliceMipmaps.length; n++) {
                    mipmaps[i] = mipmaps[i] || [];
                    mipmaps[i].push(sliceMipmaps[n][i]);
                }
            }
        } else if (value.length > 1) {
            for (let i = 0; i < value.length; ++i) {
                const slices = value[i];
                for (let n = 0; n < slices.length; ++n) {
                    mipmaps[i] = mipmaps[i] || [];
                    mipmaps[i].push(slices[n].extractMipmap0());
                }
            }
        }

        this._setMipmapParams(mipmaps);
    }

    private _setMipmapParams(value: ImageAsset[][]): void {
        this._generatedMipmaps = value;
        this._setMipmapLevel(this._generatedMipmaps.length);
        if (this._generatedMipmaps.length > 0 && this._generatedMipmaps[0]?.length > 0) {
            const imageAsset: ImageAsset = this._generatedMipmaps[0][0];
            this.reset({
                width: imageAsset.width,
                height: imageAsset.height,
                format: imageAsset.format,
                mipmapLevel: this._generatedMipmaps.length,
                baseLevel: this._baseLevel,
                maxLevel: this._maxLevel,
            });
            this._generatedMipmaps.forEach((mipmapSlices, level) => {
                for (let i = 0; i < mipmapSlices.length; ++i) {
                    this._assignImage(mipmapSlices[i], level, i);
                }
            });
        } else {
            this.reset({
                width: 0,
                height: 0,
                mipmapLevel: this._generatedMipmaps.length,
                baseLevel: this._baseLevel,
                maxLevel: this._maxLevel
            });
        }
    }

    get slicesCount(): number {
        return this._mipmaps.length > 0 && this._mipmaps[0]?.length > 0 ? this._mipmaps[0].length : 0;
    }

    get image(): ITexture2DArrayMipmaps | null {
        return this._mipmaps.length === 0 ? null : this._mipmaps[0];
    }

    set image(value) {
        this.mipmaps = value ? [value] : [];
    }

    public static create(textures: Texture2D[], slicesCount: number = 1, out?: Texture2DArray): Texture2DArray {
        const mipmaps: ITexture2DArrayMipmaps[] = [];
        const nMipmaps = textures.length / slicesCount;

        for (let i = 0; i < nMipmaps; i++) {
            const x = i * slicesCount;
            mipmaps[i].push(textures[x].image!);
        }

        out = out || new Texture2DArray();
        out.mipmaps = mipmaps;
        return out;
    }

    public initialize(): void {
        this.mipmaps = this._mipmaps;
    }

    public onLoaded(): void {
        this.initialize();
    }

    public reset(info: ITexture2DArrayCreateInfo): void {
        this._width = info.width;
        this._height = info.height;
        this._setGFXFormat(info.format);
        const mipLevels = info.mipmapLevel === undefined ? 1 : info.mipmapLevel;
        this._setMipmapLevel(mipLevels);
        const minLod = info.baseLevel === undefined ? 0 : info.baseLevel;
        const maxLod = info.maxLevel === undefined ? 1000 : info.maxLevel;
        this._setMipRange(minLod, maxLod);
        this._tryReset();
    }

    public updateMipmaps(firstLevel: number = 0, count?: number | undefined): void {
        if (firstLevel >= this._generatedMipmaps.length)
            return;

        const nUpdate = Math.min(
            count === undefined ? this._generatedMipmaps.length : count,
            this._generatedMipmaps.length - firstLevel,
        );

        for (let i = 0; i < nUpdate; i++) {
            const level = firstLevel + i;
            const slices: ImageAsset[] = this._generatedMipmaps[level];
            for (let n = 0; n < slices.length; n++) {
                this._assignImage(slices[n], level, n);
            }
        }
    }

    public description(): string {
        return `<cc.Texture2DArray | SlicesCount = ${this.slicesCount} | Dimension = ${this.width}x${this.height}>`;
    }

    public destroy(): boolean {
        this._mipmaps = [];
        this._generatedMipmaps = [];
        return super.destroy();
    }

    public releaseTexture(): void {
        this.destroy();
    }

    public _serialize(ctxForExporting: any): Record<string, unknown> | null {
        // console.warn(`Texture2DArray:_serialize`);
        if (EDITOR || TEST) {
            return {
                base: super._serialize(ctxForExporting),
                slice: this._mipmaps.length && this._mipmaps[0]?.length > 0 ? this._mipmaps[0].length : 0,
                mipmaps: this._mipmaps.map((mipmapSlices) => {
                    const output: string[] = [];
                    if (ctxForExporting && ctxForExporting._compressUuid) {
                        for (let i = 0; i < mipmapSlices.length; i++) {
                            output[i] = EditorExtends.UuidUtils.compressUuid(mipmapSlices[i]._uuid, true);
                        }
                    }
                    else {
                        for (let i = 0; i < mipmapSlices.length; i++) {
                            output[i] = mipmapSlices[i]._uuid;
                        }
                    }
                    return output;
                }),
            };
        }
        return null;
    }

    public _deserialize(serializedData: ITexture2DArraySerializeData, handle: any): void {
        // console.warn(`Texture2DArray:_deserialize`);
        const data = serializedData;
        super._deserialize(data.base, handle);

        this._mipmaps = new Array(data.mipmaps.length);
        for (let i = 0; i < data.mipmaps.length; i++) {
            this._mipmaps[i] = [];
            const mipmapSlices = data.mipmaps[i];
            const imageAssetClassId = js.getClassId(ImageAsset);

            for (let n = 0; n < data.slice; n++) {
                handle.result.push(this._mipmaps[i], `${n}`, mipmapSlices[n], imageAssetClassId);
            }
        }
    }

    protected _getGfxTextureCreateInfo(presumed: PresumedGFXTextureInfo): TextureInfo | null {
        const texInfo = new TextureInfo(TextureType.TEX2D_ARRAY);
        texInfo.width = this._width;
        texInfo.height = this._height;
        texInfo.layerCount = this.slicesCount;
        Object.assign(texInfo, presumed);
        return texInfo;
    }

    protected _getGfxTextureViewCreateInfo(presumed: PresumedGFXTextureViewInfo): TextureViewInfo | null {
        const texViewInfo = new TextureViewInfo();
        texViewInfo.type = TextureType.TEX2D_ARRAY;
        texViewInfo.baseLayer = 0;
        texViewInfo.layerCount = this.slicesCount;
        Object.assign(texViewInfo, presumed);
        return texViewInfo;
    }

    public initDefault(uuid?: string): void {
        super.initDefault(uuid);

        const imageAsset = new ImageAsset();
        imageAsset.initDefault();
        this.mipmaps = [[imageAsset]];
    }

    public validate(): boolean {
        return this.mipmaps && this.mipmaps.length > 0;
    }

}

cclegacy.Texture2DArray = Texture2DArray;

interface ITexture2DArraySerializeData {
    base: string;
    slice: number;
    mipmaps: ImageAsset[][];
}

File: native/cocos/core/assets/Texture2DArray.cpp
#include "core/assets/Texture2DArray.h"
#include "core/assets/ImageAsset.h"
#include "core/assets/Texture2D.h"
#include "renderer/gfx-base/GFXTexture.h"

namespace cc {

Texture2DArray *Texture2DArray::create(const ccstd::vector<Texture2D *> &textures, uint32_t slicesCount) {
    size_t nMipmaps = textures.size() / slicesCount;
    ccstd::vector<ITexture2DArrayMipmaps> mipmaps;
    mipmaps.reserve(nMipmaps);

    for (size_t i = 0; i < nMipmaps; i++) {
        size_t x = i * slicesCount;
        mipmaps[i].slices.emplace_back(textures[x]->getImage());
    }

    auto *out = ccnew Texture2DArray();
    out->setMipmaps(mipmaps);

    return out;
}

Texture2DArray::Texture2DArray() = default;
Texture2DArray::~Texture2DArray() = default;

void Texture2DArray::setMipmaps(const ccstd::vector<ITexture2DArrayMipmaps> &value) {
    // CC_LOG_WARNING("Texture2DArray::setMipmaps, value size: %d", value.size());
    _mipmaps = value;

    ccstd::vector<ITexture2DArrayMipmaps> mipmaps;
    mipmaps.resize(value.size());

    if (value.size() == 1) {
        const ITexture2DArrayMipmaps &mipmapsSlices = value[0];
        ccstd::vector<ITexture2DArrayMipmaps> sliceMipmaps;
        size_t maxMipmapLevel = 0;

        size_t mipmapsSlicesCnt = mipmapsSlices.slices.size();
        // CC_LOG_WARNING("Texture2DArray::setMipmaps -> mipmapsSlices.slices.size(): %d", mipmapsSlicesCnt);

        sliceMipmaps.resize(mipmapsSlicesCnt);

        for (size_t n = 0; n < mipmapsSlicesCnt; n++) {
            // CC_LOG_INFO("Extract Mipmaps at slice: %d", n);
            sliceMipmaps[n].slices = mipmapsSlices.slices[n]->extractMipmaps();
            // sliceMipmaps[n].slices.assign(extractMipmaps.begin(), extractMipmaps.end());
            maxMipmapLevel = std::max(maxMipmapLevel, sliceMipmaps[n].slices.size());
        }

        for (size_t n = 0; n < mipmapsSlicesCnt; n++) {
            if (sliceMipmaps[n].slices.size() != maxMipmapLevel) {
                assert("The number of mipmaps of each slice is different.");
                setMipmapParams({});
                return;
            }
        }

        mipmaps.resize(maxMipmapLevel);
        // CC_LOG_INFO("mipmaps resize to: %d", maxMipmapLevel);

        for (size_t i = 0; i < maxMipmapLevel; i++) {
            for (size_t n = 0; n < sliceMipmaps.size(); n++) {
                // CC_LOG_INFO("mipmaps[%d].slices.emplace_back(sliceMipmaps[%d].slices[%d])", i, n, i);
                mipmaps[i].slices.emplace_back(sliceMipmaps[n].slices[i]);
            }
        }
    } else if (value.size() > 1) {
        for (size_t i = 0; i < value.size(); i++) {
            const auto &slices = value.at(i);
            for (const auto & slice : slices.slices) {
                mipmaps[i].slices.emplace_back(slice->extractMipmap0());
            }
        }
    }

    setMipmapParams(mipmaps);
}

void Texture2DArray::setMipmapParams(const ccstd::vector<ITexture2DArrayMipmaps> &value) {
    // CC_LOG_WARNING("Texture2DArray::setMipmapParams, value.size(): %d", value.size());
    _generatedMipmaps = value;
    setMipmapLevel(static_cast<uint32_t>(_generatedMipmaps.size()));

    if (!_generatedMipmaps.empty() && !_generatedMipmaps[0].slices.empty()) {
        ImageAsset *imageAsset = _generatedMipmaps[0].slices[0];
        // CC_LOG_INFO("Reset with first generatedMipmaps represent as: _generatedMipmaps[0].slices[0], %p", imageAsset);
        reset({imageAsset->getWidth(),
               imageAsset->getHeight(),
               imageAsset->getFormat(),
               static_cast<uint32_t>(_generatedMipmaps.size()),
               _baseLevel,
               _maxLevel});
        for (size_t i = 0; i < _generatedMipmaps.size(); i++) {
            auto mipmapSlices = _generatedMipmaps[i];
            for (size_t n = 0; n < mipmapSlices.slices.size(); n++) {
                // CC_LOG_INFO("Assign image: %p, level: %d, arrayIndex: %d", mipmapSlices.slices[n].get(), i, n);
                assignImage(mipmapSlices.slices[n], i, n);
            }
        }
    } else {
        reset({0, 0, ccstd::nullopt, static_cast<uint32_t>(_generatedMipmaps.size()), _baseLevel, _maxLevel});
    }
}

void Texture2DArray::setMipmapsForJS(const ccstd::vector<ITexture2DArrayMipmaps> &value) {
    _mipmaps = value;
}

void Texture2DArray::setImage(const ITexture2DArrayMipmaps *value) {
    if (value != nullptr) {
        setMipmaps({*value});
    } else {
        setMipmaps({});
    }
}

void Texture2DArray::reset(const ITexture2DArrayCreateInfo &info) {
    _width = info.width;
    _height = info.height;
    setGFXFormat(info.format);

    uint32_t mipLevels = info.mipmapLevel.has_value() ? info.mipmapLevel.value() : 1;
    setMipmapLevel(mipLevels);

    uint32_t minLod = info.baseLevel.has_value() ? info.mipmapLevel.value() : 0;
    uint32_t maxLod = info.maxLevel.has_value() ? info.maxLevel.value() : 1000;
    setMipRange(minLod, maxLod);

    tryReset();
}

void Texture2DArray::updateMipmaps(uint32_t firstLevel, uint32_t count) {
    if (firstLevel >= _generatedMipmaps.size()) {
        return;
    }

    auto nUpdate = static_cast<uint32_t>(std::min(count == 0 ? _generatedMipmaps.size() : count, _generatedMipmaps.size() - firstLevel));

    for (size_t i = 0; i < nUpdate; i++) {
        size_t level = firstLevel + i;
        const ccstd::vector<IntrusivePtr<ImageAsset>> &slices = _generatedMipmaps[level].slices;
        for (size_t n = 0; n < slices.size(); n++) {
            assignImage(slices[n], level, n);
        }
    }
}

void Texture2DArray::initialize() {
    setMipmaps(_mipmaps);
}

void Texture2DArray::onLoaded() {
    initialize();
}

bool Texture2DArray::destroy() {
    _mipmaps.clear();
    _generatedMipmaps.clear();
    return Super::destroy();
}

void Texture2DArray::releaseTexture() {
    destroy();
}

ccstd::any Texture2DArray::serialize(const ccstd::any &ctxForExporting) {
    // FIXME: serialize
    return nullptr;
}

void Texture2DArray::deserialize(const ccstd::any &serializedData, const ccstd::any &handle) {
    const auto *data = ccstd::any_cast<Texture2DArraySerializeData>(&serializedData);
    if (data == nullptr) {
        return;
    }

    Super::deserialize(data->base, handle);
    _mipmaps.resize(data->mipmaps.size());

    for (size_t i = 0; i < data->mipmaps.size(); i++) {
        // const auto mipmapSlices = data->mipmaps[i];
        _mipmaps[i].slices.resize(data->slice);
        for (size_t n = 0; n < data->slice; n++) {
            _mipmaps[i].slices.emplace_back(ccnew ImageAsset());
        }
    }
}

gfx::TextureInfo Texture2DArray::getGfxTextureCreateInfo(gfx::TextureUsageBit usage, gfx::Format format, uint32_t levelCount, gfx::TextureFlagBit flags) {
    // CC_LOG_DEBUG("Texture2DArray::getGfxTextureCreateInfo");
    gfx::TextureInfo texInfo;
    texInfo.type = gfx::TextureType::TEX2D_ARRAY;
    texInfo.width = _width;
    texInfo.height = _height;
    texInfo.layerCount = getSlicesCount();
    texInfo.usage = usage;
    texInfo.format = format;
    texInfo.levelCount = levelCount;
    texInfo.flags = flags;
    return texInfo;
}

gfx::TextureViewInfo Texture2DArray::getGfxTextureViewCreateInfo(gfx::Texture *texture, gfx::Format format, uint32_t baseLevel, uint32_t levelCount) {
    // CC_LOG_DEBUG("Texture2DArray::getGfxTextureViewCreateInfo");
    gfx::TextureViewInfo texViewInfo;
    texViewInfo.type = gfx::TextureType::TEX2D_ARRAY;
    texViewInfo.baseLayer = 0;
    texViewInfo.layerCount = getSlicesCount();
    texViewInfo.texture = texture;
    texViewInfo.format = format;
    texViewInfo.baseLevel = baseLevel;
    texViewInfo.levelCount = levelCount;

    return texViewInfo;
}

void Texture2DArray::initDefault(const ccstd::optional<ccstd::string> &uuid) {
    Super::initDefault(uuid);

    auto *imageAsset = ccnew ImageAsset();
    imageAsset->initDefault(ccstd::nullopt);

    ITexture2DArrayMipmaps mipmap;

    mipmap.slices.emplace_back(imageAsset);
    setMipmaps({mipmap});
}

bool Texture2DArray::validate() const {
    auto mipmaps = getMipmaps();
    return !mipmaps.empty();
}

} // namespace cc
